### ISSUE TITLE
Higer dimension for layernorm

### ISSUE THREAD

**Issue by junjie18** [HazyResearch/flash-attention#295]
Hi, Thanks for your great work. 
Can you privide [layernorm](https://github.com/HazyResearch/flash-attention/tree/main/csrc/layer_norm) implementation for higher dimension, like 12288.

**Comment by tridao** [comment]
I don't have bandwidth for that right now. Hope to do that in the future.
One possible concern is that it increases compilation time.

**Comment by junjie18** [comment]
Thanks for your quick reply.

I noticed the main difference of each dimension is in`REGISTER_PARALLEL_FWD_LAUNCHER` and `REGISTER_BWD_LAUNCHER`. 

Could you kindly give me some guidance in finetuning these hyperparameters?

**Comment by tridao** [comment]
The hyperparameters determine how many threads per thread block, and how many elements is processed per thread.
They're tuned by experimentation: try out a few sensible values and see what's faster. One has to be careful to avoid register spilling.

For dimension > 8k (especially the backward pass) the best implementation would require communicating between different thread blocks. That part isn't implemented (correctly) right now, that's why it might take some work to get best performance for dim > 8k.

Btw you can also look at Apex's [LayerNorm](https://github.com/NVIDIA/apex/blob/master/apex/contrib/layer_norm/layer_norm.py) (which my implementation is based on). They have layer norm for a lot of dimensions (but no dropout or residual).

**Comment by junjie18** [comment]
Thanks for you detailed reply. 
`That part isn't implemented (correctly) right now, that's why it might take some work to get best performance for dim > 8k.` This mean the thread num in each block maybe insufficient?

**Comment by tridao** [comment]
You might get a lot of register spilling.

**Comment by junjie18** [comment]
@tridao 
Sorry to bother again. Is there any way to easily support `bfloat16` for `LN`?

**Comment by casper-hansen** [comment]
Hi @tridao, Mistral released their 7B model which now beats many Llama models, in some cases the 70B on some benchmarks. The Mistral model has a sequence length of 32k. Now is the right time to support 16k and 32k sequence lengths as that would enable approximately 15-20% faster training of models according to my own testing.

**Comment by tridao** [comment]
That has always been supported, there's no limit on sequence length.
The layer norm (which is a separate extension from FlashAttention) supports *hidden dimension* up to 8k.

**Comment by casper-hansen** [comment]
I must have misread *hidden dimension* as sequence length somehow... thanks for clarifying!
